### ISSUE TITLE
Inconsistent behaviour when comparing measurements.

### ISSUE THREAD

**Issue by zjuhasz** [unitsofmeasurement/uom-se#129]
This can actually sort of be broken down into two separate but coupled issues.

1. The `compareTo()` method of measurement is not type safe because it attempts to access a method (also `compareTo()`) that isn't guaranteed to be present. This is because a `Measurement<Q>` contains a `Quantity<Q>` not a `ComparableQuantity<Q>`. This is addressed by an instanceOf() check but if the quantity is not an instance of `ComparableQuantity` which is entirely possible if someone is trying to interop with another implementation of the base library, the measurements compareTo() method returns 0 which states that it is equal to what it is being compared to but that is most likely not actually the case. Here is the `compareTo()` method from AbstractMeasurement:

```
    @SuppressWarnings("unchecked")
    @Override
    public int compareTo(Object o) {
      if (o instanceof Measurement) {
        Measurement m = (Measurement) o;
        if (getQuantity() instanceof ComparableQuantity) {
          ComparableQuantity<?> comp = (ComparableQuantity<?>) getQuantity();
          return comp.compareTo(m.getQuantity()) + getInstant().compareTo(m.getInstant());
        }
        return 0;
      }
      return 0;
}
```

As far as I can tell getting a quantity from a Measurement is the only way to get something that isn't guaranteed to implement ComparableQuantity, which can force users to either do constant type checks and error handling or unsafely cast Quantites to ComparableQuantities if they want measurements to play nice with the rest of UOM (which generally returns ComparableQuantities). I think there are several decent options to fix this. My preference would be to have Quantity itself implement Comparable, is there a reason why it doesn't? If that won't work I think the `getQuantity()` method of Measurement should return a ComparableQuantity not a plain old Quantity. 

These two are vastly preferred but if neither is possible (I don't see why they wouldn't be) I think it's preferable to throw a (runtime? exception) if the `instanceOf()` check fails than to  return 0. Returning 0 completely breaks the guarantees we expect of Comparable and could produce very strange and hard to find bugs in implementing code. Although like I said, I don't think this is a real solution, just slightly preferable to returning 0.

2. This leads to the second issue, fixing this basically eliminates the need to do much of what I said. I don't think Measurement can implement Comparable in a way that is consistent with the functionality expected from the Comparable interface. 

The Javadoc states:
"It is strongly recommended (though not required) that natural orderings be consistent with equals. This is so because sorted sets (and sorted maps) without explicit comparators behave "strangely" when they are used with elements (or keys) whose natural ordering is inconsistent with equals. In particular, such a sorted set (or sorted map) violates the general contract for set (or map), which is defined in terms of the equals method."

Although I don't think equals is currently implemented for measurement I would think it should check both quantity and time, whereas its `compareTo()` method completely ignores the time at which the measurement is taken and just pushes that call through to one of its members in an unsafe way. This forces it to be inconsistent with the proper implementation of `equals()`.  I don't think it has a significant impact on implementing code to have to say `measurement.getQuantity().compareTo(otherMes.getQuantity())` as opposed to measurement.compareTo(otherMeasurement) which at current would theoretically produce identical results. But this takes us back to the first issue of measurement not returning ComparableQuantity so that last code snippet I gave wouldn't be possible at present without all sorts of messy unsafe casting.

In summary I think the best solution is to make Quantity implement Comparable and to make Measurement not implement Comparable and probably to override equals() in Measurement or AbstractMeasurement since unlike compareTo() equals() can have consistent behavior in Measurement. But like I mentioned above I can see some variations of these options as well. I'd be happy to make a pull request with my suggested changes if you'd like me to @keilw

**Comment by keilw** [comment]
I'm afraid Quantity cannot extend Comparable. We had this in the JIRA of JSR 363 and it was decided to leave this implementation specific. The JDK also backs this e.g. in case of Number. concrete classes like BigDecimal implement Comparable, but Number itself doen't.

Looking at a Quantity that may not be a child of AbstractQuantity or implement ComparableQuantity is a good idea. The 1.0.1 release of uom-se is on its way, so thanks for the input, will try to get that into the upcoming release.

**Comment by keilw** [comment]
If you have a PR that does not rely on Quantity itself extending Comparable, happy if you could submit it.

**Comment by zjuhasz** [comment]
Submitted pull request #130 that hopefully strikes a good balance.
